trap.go

```diff
@@ -207,57 +228,90 @@ func (t *TrapListener) listenUDP(addr string) error {
+			}
+			if trap.Version == Version3 && trap.SecurityModel == UserSecurityModel && t.Params.SecurityModel == UserSecurityModel {
+				securityParams, ok := t.Params.SecurityParameters.(*UsmSecurityParameters)
+				if !ok {
+					t.Params.Logger.Printf("TrapListener: Invalid SecurityParameters types")
+				}
+				packetSecurityParams, ok := trap.SecurityParameters.(*UsmSecurityParameters)
+				if !ok {
+					t.Params.Logger.Printf("TrapListener: Invalid SecurityParameters types")
+				}
+				snmpEngineID := securityParams.AuthoritativeEngineID
+				msgAuthoritativeEngineID := packetSecurityParams.AuthoritativeEngineID
+				if msgAuthoritativeEngineID != snmpEngineID {
+					if 4 < len(msgAuthoritativeEngineID) && len(msgAuthoritativeEngineID) < 32 {
+						// RFC3414 3.2.3a: continue processing
+					} else {
+						err := t.reportAuthoritativeEngineID(trap, snmpEngineID, remote)
+						if err != nil {
+							t.Params.Logger.Printf("TrapListener: %s\n", err)
+						}
+						continue
 					}
 				}
 			}
+			// Here we assume that t.OnNewTrap will not alter the contents
+			// of the PDU (per documentation, because Go does not have
+			// compile-time const checking).  We don't pass a copy because
+			// the SnmpPacket type is somewhat large, but we could without
+			// violating any implicit or explicit spec.
+			t.OnNewTrap(trap, remote)
+
+			// If it was an Inform request, we need to send a response.
+			if trap.PDUType == InformRequest { //nolint:whitespace
+
+				// Reuse the packet, since we're supposed to send it back
+				// with the exact same variables unless there's an error.
+				// Change the PDUType to the response, though.
+				trap.PDUType = GetResponse
+
+				// If the response can be sent, the error-status is
+				// supposed to be set to noError and the error-index set to
+				// zero.
+				trap.Error = NoError
+				trap.ErrorIndex = 0
+
+				// TODO: Check that the message marshalled is not too large
+				// for the originator to accept and if so, send a tooBig
+				// error PDU per RFC3416 section 4.2.7.  This maximum size,
+				// however, does not have a well-defined mechanism in the
+				// RFC other than using the path MTU (which is difficult to
+				// determine), so it's left to future implementations.
+				err := t.SendUDP(trap, remote)
+				if err != nil {
+					t.Params.Logger.Printf("TrapListener: %s\n", err)
+				}
+			}
 		}
 	}
 }
 
+func (t *TrapListener) reportAuthoritativeEngineID(trap *SnmpPacket, snmpEngineID string, addr *net.UDPAddr) error {
+	newSecurityParams, ok := trap.SecurityParameters.Copy().(*UsmSecurityParameters)
+	if !ok {
+		return errors.New("unable to cast SecurityParams to UsmSecurityParameters")
+	}
+	newSecurityParams.AuthoritativeEngineID = snmpEngineID
+	reportPacket := trap
+	reportPacket.PDUType = Report
+	reportPacket.MsgFlags &= AuthPriv
+	reportPacket.SecurityParameters = newSecurityParams
+	reportPacket.Variables = []SnmpPDU{
+		{
+			Name:  ".1.3.6.1.6.3.15.1.1.4.0",
```

Comment:
I think it can help to have a constant for this OID usmStatsUnknownEngineIDs, it will help readability.

Or alternatively a comment indicating the OID correspond to `usmStatsUnknownEngineIDs`.